minerl/herobraine/hero/handlers/agent/actions/equip.py

```diff
@@ -1,45 +1,59 @@
-import jinja2
-import minerl.herobraine.hero.spaces as spaces
+import logging
 
+from minerl.herobraine.hero.handlers.agent import action
+from minerl.herobraine.hero.handlers import util
+import minerl.herobraine.hero.mc as mc
 
-class EquipAction(ItemListAction):
+
+class EquipAction(action.ItemWithMetadataListAction):
     """
     An action handler for observing a list of equipped items
     """
 
-    def to_string(self):
-        return 'equip'
-
-    def xml_template(self) -> str:
-        return str("<EquipCommands/>")
+    logger = logging.getLogger(__name__ + ".EquipAction")
 
     def __init__(self, items: list, _default='none', _other='other'):
         """
         Initializes the space of the handler to be one for each item in the list plus one for the
         default no-craft action
         """
-        self._items = items
-        self._command = 'equip'
-        super().__init__(self._command, self._items, _default=_default, _other=_other),
+        super().__init__("equip", items, _default=_default, _other=_other)
         self.previous = self._default
 
-    def from_universal(self, obs):
-        try:
-            if obs['slots']['gui']['type'] == 'class net.minecraft.inventory.ContainerPlayer':
-                hotbar_index = int(obs['hotbar'])
-                item = self._univ_items.index(obs['slots']['gui']['slots'][-10 + hotbar_index]['name'])
-                if item != self.previous:
-                    self.previous = item
-                    return obs['slots']['gui']['slots'][-10 + hotbar_index]['name'].split('minecraft:')[-1]
-        except KeyError:
-            return self._default
-        except ValueError:
-            return self._other
-        return self._default
+    def xml_template(self) -> str:
+        return str("<EquipCommands/>")
 
     def reset(self):
         self.previous = self._default
+
+    def from_universal(self, obs) -> str:
+        slots_gui_type = obs['slots']['gui']['type']
+        if slots_gui_type == 'class net.minecraft.inventory.ContainerPlayer':
+            hotbar_index = int(obs['hotbar'])
+            hotbar_slot = obs['slots']['gui']['slots'][-10 + hotbar_index]
+
+            item_type = mc.strip_item_prefix(hotbar_slot['name'])
+            metadata = hotbar_slot['variant']
+            id = util.get_unique_matching_item_list_id(self.items, item_type, metadata,
+                                                       clobber_logs=False)
+            if id is None:
+                id = "other"
+
+            # TODO(shwang): Tell Brandon that this is a change in behavior (probably ok?)
+            #   Consecutive "other" now results in "none". It used to be the case that
+            #   "other" would always be returned regardless of repetition and whether
+            #   ground-truth-"other" equipped items were being changed. This behavior
+            #   was different from the behavior of other keys.
+            if id == self.previous:
```

Comment:
This is a good point - neither are Idea, the best case here would be if item @ t=1 != item @ t=2 iff we see a non "none" equip message, such that consecutive others represent changing to different items in the inventory